<?xml version="1.0"?>
<launch>
   <!-- Global Arguments -->
   <arg name="robot_name" default="locobot"/>
   <arg name="implementation" default="cpp"/>  <!-- Options: "python" or "cpp" -->
   <arg name="test_duration" default="60"/>   <!-- Test duration in seconds -->
   
   <!-- Camera Topics -->
   <arg name="rgb_topic" default="/$(arg robot_name)/camera/color/image_raw"/>
   <arg name="annotated_image_topic" default="camera/yolo/annotated_image"/>
   
   <!-- Model paths for both implementations -->
   <arg name="python_model_path" default="$(find yolo_vision)/models/best.pt"/>
   <arg name="cpp_model_path" default="$(find yolo_vision)/models/best.onnx"/>
   
   <!-- Launch the appropriate YOLO node based on implementation -->
   <group if="$(eval implementation == 'python')">
     <node name="yolo_node" pkg="yolo_vision" type="yolo_node.py" output="screen" ns="$(arg robot_name)">
         <param name="model_path" value="$(arg python_model_path)"/>
         <param name="image_topic" value="$(arg rgb_topic)"/>
         <param name="annotated_image_topic" value="$(arg annotated_image_topic)"/>
     </node>
   </group>
   
   <group if="$(eval implementation == 'cpp')">
     <node name="yolo_node" pkg="yolo_vision" type="yolo_node" output="screen" ns="$(arg robot_name)">
         <param name="model_path" value="$(arg cpp_model_path)"/>
         <param name="image_topic" value="$(arg rgb_topic)"/>
         <param name="annotated_image_topic" value="$(arg annotated_image_topic)"/>
         <param name="conf_threshold" value="0.5"/>
         <param name="nms_threshold" value="0.4"/>
         <param name="max_detections" value="100"/>
         <param name="process_rate" value="30.0"/>
     </node>
   </group>
   
   <!-- Launch the performance tester -->
   <node name="yolo_performance_tester" pkg="yolo_vision" type="yolo_performance_tester.py" output="screen">
     <param name="implementation" value="$(arg implementation)"/>
     <param name="test_duration" value="$(arg test_duration)"/>
     <param name="image_topic" value="$(arg rgb_topic)"/>
     <param name="annotated_topic" value="/$(arg robot_name)/$(arg annotated_image_topic)"/>
   </node>
</launch>